aggregator.go

```diff
@@ -0,0 +1,139 @@
+	"log"
+	"os"
+	"sync"
+	"time"
+)
+
+func getLatestOutput(instanceId string, sess *session.Session, withLatest bool) string {
+	var consoleInput *ec2.GetConsoleOutputInput
+	if withLatest {
+		consoleInput = &ec2.GetConsoleOutputInput{
+			InstanceId: aws.String(instanceId),
+			Latest:     aws.Bool(true),
+		}
+	} else {
+		consoleInput = &ec2.GetConsoleOutputInput{
+			InstanceId: aws.String(instanceId),
+		}
+	}
+	ec2Svc := ec2.New(sess)
+
+	result, err := ec2Svc.GetConsoleOutput(consoleInput)
+	if err != nil {
+		if err, ok := err.(awserr.Error); ok {
+			// Non Nitro AWS instance types don't support retrieving the latest console logs
+			if err.Code() == "UnsupportedOperation" && withLatest {
+				return getLatestOutput(instanceId, sess, false)
+			} else {
+				log.Println(err.Error())
+			}
+		} else {
+			log.Println(err.Error())
+		}
+		return ""
+	}
+
+	output, err := base64.StdEncoding.DecodeString(*result.Output)
+	if err != nil {
+		log.Printf("Failed decoding console output from base64\n%s\n", err)
+		return ""
+	}
+
+	return string(output)
+}
+
+func findOverlappingEndingIndex(prev, new string) int {
+	var i int
+	prevLen := len(prev)
+	for i = len(new); i > 0; i-- {
+		if i > prevLen {
+			continue
+		}
+		if new[:i] == prev[prevLen-i:] {
+			break
+		}
+	}
+	return i
+}
+
+func aggregateConsoleOutput(wg *sync.WaitGroup, folder, instanceId string, sess *session.Session) {
+	defer wg.Done()
+
+	logPath := fmt.Sprintf("%s/%s.log", folder, instanceId)
+	f, err := os.OpenFile(logPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		log.Printf("Failed opening file: %s\n", logPath)
+		log.Fatalln(err)
+	}
+	defer f.Close()
+
+	prev := ""
+	total := 0
+	for {
+		output := getLatestOutput(instanceId, sess, true)
+		ind := findOverlappingEndingIndex(prev, output)
+		newData := output[ind:]
+		if newData != "" {
+			log.Printf("Instance %s got new console output, updating log file %s\n", instanceId, logPath)
+			if _, err := f.WriteString(newData + "\n"); err != nil {
+				log.Printf("Failed writing new console output to file\n%s\n", err)
+			} else {
+				total += len(newData)
+				log.Printf("Total characters written to log file so far: %d\n", total)
+			}
+		} else {
+			log.Printf("Instance %s no new console output\n", instanceId)
+		}
+
+		prev = output
+		time.Sleep(5 * time.Minute)
+
+	}
+}
+
+type arrayFlag []string
+
+func (i *arrayFlag) String() string {
+	return "my string representation"
+}
+
+func (i *arrayFlag) Set(value string) error {
+	*i = append(*i, value)
+	return nil
+}
+
+func main() {
+	log.Println("Running aws console log aggregator service...")
+	var ids arrayFlag
+
+	region := flag.String("region", "eu-central-1", "AWS instance region")
+	folder := flag.String("folder", ".", "Aggregated AWS console log folder")
+	flag.Var(&ids, "id", "List of AWS instances ids")
+	flag.Parse()
+	log.Printf("Region:%s OutputFolder:'%s' InstancesIds:%s\n", *region, *folder, ids)
```

Comment:
No need for a newline at the end of each log line: https://golang.org/pkg/log/#Logger.Output